### PR TITLE
Fix 0115A arena for manipulation

### PR DESCRIPTION
**Pull request by yabjames**
Closes TheRARELab/fetchit#13
I used the fetchit tables and poses of the gearboxes and caddies from the montreal2019 world. 

https://github.com/user-attachments/assets/65ea7950-1adb-4dae-a4ed-16e3b2037031

https://github.com/user-attachments/assets/adade585-2027-4ca1-97a8-c3145da9bd4a

### CODE REVIEW

**Comment by umhan35** [comment]
Oops. I just saw this while checking the project board. I actually didn't receive any email notification for this because I am not a reviewer. I just added myself and Xiangfei there.

Could you review our [workflow policy](https://usfedu.sharepoint.com/:w:/r/teams/RARELabGRP/Shared%20Documents/General/Lab%20handbook/Workflow/Workflow%20Policy.docx?d=wa2499df3316c473eb97b783775fd2308&csf=1&web=1&e=zgkBlR)?

Thanks!

File: fetchit_challenge/worlds/nec0115a_arena.world [hunk 1]
```diff
@@ -14,58 +14,57 @@
     </include>
 
     <include>
-      <uri>model://rarelab_long_table</uri>
-      <name>gearbox_table</name>
-      <pose>0.468304 1.457626 0.0 0 0 -3.141585</pose>
+      <uri>model://fetchit_table_fetch</uri>
+      <name>fetchit_table_caddy</name>
+      <pose>-0.135 -1.28849 0.0 0 0 3.1416</pose>
```
Comment (umhan35) [inline fetchit_challenge/worlds/nec0115a_arena.world]: We do need to use the long table that is in the NEC arena here.

```suggestion
      <uri>model://rarelab_long_table</uri>
      <name>gearbox_table</name>
      <pose>0.468304 1.457626 0.0 0 0 -3.141585</pose>
```

File: fetchit_challenge/worlds/nec0115a_arena.world [hunk 2]
```diff
@@ -14,58 +14,57 @@
...
     </include>
 
-    <include>
-      <uri>model://rarelab_long_table</uri>
-      <name>caddy_table</name>
-      <pose>0.710161 -0.806812 0.0 0 0 -3.141585</pose>
-    </include>
-
-
```
Comment (umhan35) [inline fetchit_challenge/worlds/nec0115a_arena.world]: Need to revert this change as well.

File: fetchit_challenge/worlds/nec0115a_arena.world [hunk 3]
```diff
@@ -14,58 +14,57 @@
...
       <uri>model://caddy_yellow</uri>
       <name>caddy_yellow1</name>
-      <pose>-0.604281 -0.942377 0.646347 0 0 0</pose>
+      <pose>-0.414898 -1.20441 0.826729 0 0 0</pose>
+    </include>
+
+    <include>
+      <uri>model://fetchit_table_fetch</uri>
+      <name>fetchit_table_gearbox_bottom</name>
+      <pose>-0.305 1.225045 0.0 0 0 0</pose>
```
Comment (umhan35) [inline fetchit_challenge/worlds/nec0115a_arena.world]:
```suggestion
```

**Review by umhan35 — CHANGES_REQUESTED** [review]
We actually need to have the long tables that are in the NEC arena here. Could you add them back and test the new object poses again? Thanks!

**Comment by yabjames** [comment]
> Oops. I just saw this while checking the project board. I actually didn't receive any email notification for this because I am not a reviewer. I just added myself and Xiangfei there.
> 
> Could you review our [workflow policy](https://usfedu.sharepoint.com/:w:/r/teams/RARELabGRP/Shared%20Documents/General/Lab%20handbook/Workflow/Workflow%20Policy.docx?d=wa2499df3316c473eb97b783775fd2308&csf=1&web=1&e=zgkBlR)?
> 
> Thanks!

Okay sorry, I forgot to add you and Xiangfei as a reviewer! I'll look over the workflow policy again.

**Comment by yabjames** [comment]
I attempted to use the reverted version, using the detection using the long table and the original gearbox/caddy poses from the montreal2019 world, but it did not work. I believe the problem is that the table height is too low for the detection to work. I did a separate test where I moved the long table higher and it started to work. 

Testing with original poses and long table:
image

Test with higher floating long table:

https://github.com/user-attachments/assets/39f98921-0395-4800-aa9e-2ce225e021cb

**Comment by umhan35** [comment]
I see. Xiangfei actually [changed the robot position in front of the gearbox bottom table](https://github.com/TheRARELab/fetchit/commit/e048786005bc8cdfc2b06e53541ab00d86098288#diff-c60bd571febd0b3c69f38946d5e9c5f35687202598951f5a013b657cbc809f52) and [successfully test here](https://github.com/TheRARELab/ar_physical_ros/issues/1#issuecomment-2198230788). So you need to go to a new location. This is the new pose:

image

You also want to change the object positions. See what has been tested in the room: (https://github.com/TheRARELab/ar_physical_ros/issues/1#issuecomment-2198230788).

Let me know how it goes. @xiangfeikong777 can also help.

**Comment by umhan35** [comment]
For reference: This is the test for placing into caddy: https://github.com/TheRARELab/ar_physical_ros/issues/2

**Comment by yabjames** [comment]
I tested using the new pose. The fetch robot is not in the arena for that position. I may be misunderstanding what you meant. I have looked through the tests that Xiangfei has done and I have tested different poses with the long table, but no success.

I checked issue #1 to see the manipulation test. I am not too sure why Gazebo is not giving the same result as the real life task. I noticed that the height difference between the fetch robot and the table is different from real life and Gazebo. I think the fetch robot extends its height more in the Gazebo simulation. I also noticed that the fetch robot manipulation starts working when the gearbox bottoms are near the height of the top of the black waist portion of the robot.

Is there a way to extend the height less in Gazebo? I might check the height difference when I go in the lab.  `roscd caddy_manipulation && rviz -d moveit.rviz` does not show anything unless the table is closer. 

New Pose (1.940, 1.412, 0):
image

Height Comparisons between fetch waist and table (real life compared to Gazebo):
image
image
image
image

**Comment by umhan35** [comment]
> I tested using the new pose. The fetch robot is not in the arena for that position.

Sorry. I almost forgot it (I stopped working on fetchit3 years ago). The coordinate system is different in Gazebo: The center is the origin. Here is the map's: It is the [bottom left corner](https://github.com/TheRARELab/fetchit/blob/main/nav_config/README.md#coordinate-system).

image

> `roscd caddy_manipulation && rviz -d moveit.rviz` does not show anything unless the table is closer.

I think this is what we want. Becasue the task _was_ closer in Xiangfei's test, which is how to compare this virtual/virtual condition with the physical condition.

So we probably need a new Fetch pose to go to the front of Gazebo instead of those we used for `/gazebo/set_model_state`.

> Is there a way to extend the height less in Gazebo? I might check the height difference when I go in the lab.

 It is less likely Gazebo fetch extends less than the real one...

It is more likely that the virtual table ([folder](https://github.com/TheRARELab/fetch_gazebo/tree/gazebo9/fetchit_challenge/models/rarelab_long_table/), [meshes](https://github.com/TheRARELab/fetch_gazebo/tree/gazebo9/fetchit_challenge/models/rarelab_long_table/meshes)) is lower than our physical table.

You and Xiangfei can check it Monday.

image

**Comment by umhan35** [comment]
Ok. The virtual table is 0.61m (see below). We need to check the physical table's height (and double-check the width & length) tomorrow (Monday). I don't think the Fetch robot's extension is a problem...

image

**Comment by yabjames** [comment]
Yeah, you are definitely right about the table height. I am not sure why I thought robot extended too much. I thought the table height would be correct, but I just realized 61cm is actually near my knees! I will bring a measuring tape on Monday. Then, I can change the table height in blender and hopefully the manipulation will work. Sorry this task is taking so long!

**Comment by umhan35** [comment]
Cool. Let's figure it out Monday. Have a good weekend!

**Comment by yabjames** [comment]
@umhan35 
I made a new docs PR to show where to place the fetch robot at TheRARELab/ar_physical_ros/pull/21. Not sure if I am supposed to add the "closes" keyword to the PR since there is no relevant issue for it.
I used the difference in width between the montreal table and the `rare_long_table` to precisely change the poses of the caddy, gearbox bottom, and new fetch robot model state. I tested the manipulation and it can consistently detect the caddy compartments.

Screenshot 2024-08-06 131650
Screenshot 2024-08-06 133040

File: fetchit_challenge/worlds/nec0115a_arena.world (hunk 2 above)
Comment (yabjames) [reply fetchit_challenge/worlds/nec0115a_arena.world]: fixed in latest commit

File: fetchit_challenge/worlds/nec0115a_arena.world [hunk 4]
```diff
@@ -22,53 +22,34 @@
...
-      <pose>-0.329409 1.222056 0.690921 3.141546 0 -2.437420</pose>
-    </include>
-
-    <include>
-      <uri>model://gearbox_bottom_grey</uri>
-      <name>gearbox_bottom_grey4</name>
-      <pose>-0.482731 1.025975 0.635592 -1.570796 0 -2.472675</pose>
-    </include>
-
-
```
Comment (umhan35) [inline fetchit_challenge/worlds/nec0115a_arena.world]: Can you put back those gearbox bottoms (with working poses) for testing purposes? We can use https://github.com/TheRARELab/fetch_gazebo/blob/gazebo9/fetchit_challenge/worlds/nec0115a_arena_virtual_virtual.world for the one-gearbox-bottom case. What do you think?

**Review by umhan35 — CHANGES_REQUESTED** [review]
👍  Just one minor comment.

File: fetchit_challenge/worlds/nec0115a_arena.world (hunk 4 above)
Comment (umhan35) [reply fetchit_challenge/worlds/nec0115a_arena.world]: I.e., we will replicate it here to where we put the objects on the physical table, but just remove 1 (which is the one in the virtual/virtual world)

Comment (yabjames) [reply fetchit_challenge/worlds/nec0115a_arena.world]: okay, I put them back with the correct poses

**Comment by umhan35** [comment]
One more thing: We need to update the last image in the readme: https://github.com/TheRARELab/fetch_gazebo/tree/gazebo9. After that, it should be good to merge.

Delegating to @xiangfeikong777

**Review by umhan35 — APPROVED** [review]

**Comment by umhan35** [comment]
> One more thing: We need to update the last image in the readme: https://github.com/TheRARELab/fetch_gazebo/tree/gazebo9. After that, it should be good to merge.

@xiangfeikong777: @YingJames hasn't updated the image yet...
@umhan35  Sorry, I missed the comment you want to update the image. Can I reverse the merge?
